Fable.Remoting.MsgPack/Write.fs

```diff
@@ -400,243 +412,269 @@ let serializeObj (x: obj) (out: Stream) =
-serializerCache.Add (typeof<int16>, fun x out -> writeInt64 (x :?> int16 |> int64) out)
-serializerCache.Add (typeof<int64>, fun x out -> writeInt64 (x :?> int64) out)
-serializerCache.Add (typeof<UInt32>, fun x out -> writeUInt64 (x :?> UInt32 |> uint64) out)
-serializerCache.Add (typeof<UInt16>, fun x out -> writeUInt64 (x :?> UInt16 |> uint64) out)
-serializerCache.Add (typeof<UInt64>, fun x out -> writeUInt64 (x :?> UInt64) out)
-serializerCache.Add (typeof<float32>, fun x out -> writeSingle (x :?> float32) out)
-serializerCache.Add (typeof<float>, fun x out -> writeDouble (x :?> float) out)
-serializerCache.Add (typeof<decimal>, fun x out -> writeDouble (x :?> decimal |> float) out)
-serializerCache.Add (typeof<byte[]>, fun x out -> writeBin (x :?> byte[]) out)
-serializerCache.Add (typeof<bigint>, fun x out -> writeBin ((x :?> bigint).ToByteArray ()) out)
-serializerCache.Add (typeof<Guid>, fun x out -> writeBin ((x :?> Guid).ToByteArray ()) out)
-serializerCache.Add (typeof<DateTime>, fun x out -> writeInt64 (x :?> DateTime).Ticks out)
-serializerCache.Add (typeof<DateTimeOffset>, fun x out -> writeDateTimeOffset out (x :?> DateTimeOffset))
-serializerCache.Add (typeof<TimeSpan>, fun x out -> writeInt64 (x :?> TimeSpan).Ticks out)
+            writeObject vals.[0] types.[0] out
 
-#endif
+    and writeObject (x: obj) (t: Type) (out: ResizeArray<byte>) =
+        #if !FABLE_COMPILER
+        raise (NotSupportedException "This function is meant to be used in Fable, please use serializeObj or makeSerializer.")
+        #else
+        if isNull x then writeNil out else
+
+        match serializerCache.TryGetValue t with
+        | true, writer ->
+            writer x out
+        | _ ->
+            if FSharpType.IsRecord (t, true) then
+                let fieldTypes = FSharpType.GetRecordFields (t, true) |> Array.map (fun x -> x.PropertyType)
+                cacheGetOrAdd (t, fun x out -> writeRecord out fieldTypes (FSharpValue.GetRecordFields (x, true))) x out
+            elif t.IsArray then
+                let elementType = t.GetElementType ()
+                cacheGetOrAdd (t, fun x out -> writeArray out elementType (x :?> System.Collections.ICollection)) x out
+            elif FSharpType.IsUnion (t, true) then
+                cacheGetOrAdd (t, fun x out ->
+                    let case, fields = FSharpValue.GetUnionFields (x, t, true)
+                    let fieldTypes = case.GetFields () |> Array.map (fun x -> x.PropertyType)
+                    writeUnion out case.Tag fieldTypes fields) x out
+            elif FSharpType.IsTuple t then
+                let fieldTypes = FSharpType.GetTupleElements t
+                cacheGetOrAdd (t, fun x out -> writeTuple out fieldTypes (FSharpValue.GetTupleFields x)) x out
+            elif t.IsEnum then
+                cacheGetOrAdd (t, fun x -> writeInt64 (box x :?> int64)) x out
+            elif t.IsGenericType then
+                let tDef = t.GetGenericTypeDefinition()
```

Comment:
This is a good refactor. Might be a good idea to pull `t.GetGenericArguments ()` from the 4 branches into outer scope too.